File: app/src/main/java/com/liang/example/androidtest/MainActivity.java
package com.liang.example.androidtest;

import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.liang.example.ktandroidtest.ConstantsKt;
import com.liang.example.recyclerviewtest.recycler1.RVAdapterTest;
import com.liang.example.recyclerviewtest.recycler1.RVViewHolderTest;
import com.liang.example.utils.ApiManager;
import com.liang.example.basic_ktx.ArrayApiKt;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    public static void bindActivityList(String[] ns, String[] ds, String[] as, String[] crs, String[] us, Class[] cls, Activity a, String t) {
        ApiManager.LOGGER.d(t, "onCreate -- begin creating Activity List");
        int length = cls.length;
        List<ActivityItem> dataSet = new ArrayList<>(length);
        for (int i = 0; i < length; i++) {
            dataSet.add(new ActivityItem(ns[i], ds[i], as[i], crs[i], us[i], cls[i]));
        }

        RecyclerView rv = a.findViewById(R.id.test_activity_list);
        rv.setHasFixedSize(true);
        LinearLayoutManager activityLayoutManager = new LinearLayoutManager(a, RecyclerView.VERTICAL, false);
        activityLayoutManager.setInitialPrefetchItemCount(4);  // 数据预取(https://juejin.im/entry/58a3f4f62f301e0069908d8f)
        activityLayoutManager.setItemPrefetchEnabled(true);
        rv.setLayoutManager(activityLayoutManager);

        RVAdapterTest<ActivityItem> rvAdapter = new RVAdapterTest<ActivityItem>(dataSet, a, R.layout.item_activity_list, rv) {
            public void bindView(RVViewHolderTest viewHolder, ActivityItem data, int position) {
                viewHolder.getViewById(R.id.test_activity_item_image).setBackgroundResource(R.mipmap.ic_launcher);
                ((TextView) viewHolder.getViewById(R.id.test_activity_item_name)).setText(data.getName());
                ((TextView) viewHolder.getViewById(R.id.test_activity_item_message)).setText(
                        String.format("%s / %s / %s", data.getAuthor(), data.getCreated(), data.getUpdated()));
                ((TextView) viewHolder.getViewById(R.id.test_activity_item_desc)).setText(data.getDesc());
            }
        };
        rvAdapter.setOnItemClickListener(new RVAdapterTest.OnItemClickListener<ActivityItem>() {
            @Override
            public void onItemClick(View view, ActivityItem data, int position) {
                a.startActivity(new Intent(a, data.getClazz()));
                // a.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
                // a.overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
            }

            @Override
            public boolean onItemLongClick(View view, ActivityItem data, int position) {
                Toast.makeText(a, data.getDesc(), Toast.LENGTH_LONG).show();
                return false;
            }
        });
        // rv.addOnItemTouchListener(new RecyclerView.SimpleOnItemTouchListener() {
        //     @Override
        //     public boolean onInterceptTouchEvent(@NotNull RecyclerView rv, @NotNull MotionEvent e) {
        //         View view = rv.findChildViewUnder(e.getX(), e.getY());
        //         if (view == null) return false;
        //         // final RVViewHolderTest holder = (RVViewHolderTest) activityList.getChildViewHolder(view);
        //         int position = rv.getChildAdapterPosition(view);
        //         ActivityItem data = rvAdapter.getItem(position);
        //         a.startActivity(new Intent(a, data.getClazz()));
        //         a.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        //         a.overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
        //         return false;
        //     }
        //
        //     // private GestureDetectorCompat gestureDetectorCompat = new GestureDetectorCompat(a, new GestureDetector.SimpleOnGestureListener() {
        //     //     @Override
        //     //     public boolean onSingleTapConfirmed(MotionEvent e) {
        //     //         View view = rv.findChildViewUnder(e.getX(), e.getY());
        //     //         if (view == null) return false;
        //     //         int position = rv.getChildAdapterPosition(view);
        //     //         ActivityItem data = rvAdapter.getItem(position);
        //     //         a.startActivity(new Intent(a, data.getClazz()));
        //     //         a.overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        //     //         a.overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
        //     //         return true;
        //     //     }
        //     // });
        //     //
        //     // @Override
        //     // public boolean onInterceptTouchEvent(@NonNull RecyclerView rv, @NonNull MotionEvent e) {
        //     //     gestureDetectorCompat.onTouchEvent(e);
        //     //     return false;
        //     // }
        // });
        rv.setAdapter(rvAdapter);
        // activityList.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        ApiManager.LOGGER.d(t, "onCreate -- finish creating Activity List, ANDROID_VERSION_CODE: %d", Build.VERSION.SDK_INT);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        bindActivityList(
                ArrayApiKt.plus(Constants.names, ConstantsKt.getNames()),
                ArrayApiKt.plus(Constants.descs, ConstantsKt.getDescs()),
                ArrayApiKt.plus(Constants.authors, ConstantsKt.getAuthors()),
                ArrayApiKt.plus(Constants.created, ConstantsKt.getCreated()),
                ArrayApiKt.plus(Constants.updated, ConstantsKt.getUpdated()),
                ArrayApiKt.plus(Constants.classes, ConstantsKt.getClasses()),
                this, "App_Main");
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        TextView textView = ((ApplicationTest) getApplication()).textView;
        if (textView.getParent() != null && textView.getParent() instanceof ViewGroup) {
            ((ViewGroup) textView.getParent()).removeView(textView);
        }
        ((LinearLayout) findViewById(R.id.test_activity_root)).addView(textView, 0);
    }

    @Override
    protected void onPause() {
        super.onPause();
        ((LinearLayout) findViewById(R.id.test_activity_root)).removeViewAt(0);
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onRestart() {
        super.onRestart();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    public void onSaveInstanceState(@NotNull Bundle outState, @NotNull PersistableBundle outPersistentState) {
        super.onSaveInstanceState(outState, outPersistentState);
    }

    @Override
    protected void onRestoreInstanceState(@NotNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
    }

    @SuppressWarnings("unused")
    public void in(View v) {
        startActivity(new Intent(this, MainActivity.class));
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);  // 可以同时使用两个动画，甚至更多
    }

    @SuppressWarnings("unused")
    public void out(View v) {
        finish();
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        overridePendingTransition(android.R.anim.slide_in_left, android.R.anim.slide_out_right);
    }
}

File: app/src/main/kotlin/com/liang/example/blocktest/MainActivity.kt
package com.liang.example.blocktest

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.liang.example.androidtest.ApplicationTest
import com.liang.example.androidtest.R
import com.liang.example.utils.view.showToast

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        com.liang.example.androidtest.MainActivity.bindActivityList(names, descs, authors, created, updated, classes, this, "View_Block_Widget_Fragment")

        val textView = (application as ApplicationTest).textView
        (findViewById<View>(R.id.test_activity_root) as LinearLayout).addView(textView, 0)
        textView.setOnClickListener {
            MODE = true
            testBlockAndWidgetAndFragment(textView)
        }
    }

    private var flag = 0
    private var startCount = 0
    private val blockTotalTime = mutableListOf<Long>()
    private val blockAverageTime = mutableListOf<Long>()
    private val blockOnCreateTotalTime = mutableListOf<Long>()
    private val widgetTotalTime = mutableListOf<Long>()
    private val widgetAverageTime = mutableListOf<Long>()
    private val widgetOnCreateTotalTime = mutableListOf<Long>()
    private val fragmentTotalTime = mutableListOf<Long>()
    private val fragmentAverageTime = mutableListOf<Long>()
    private val fragmentOnCreateTotalTime = mutableListOf<Long>()

    @SuppressLint("SetTextI18n")
    private fun testBlockAndWidgetAndFragment(textView: TextView) {
        if (startCount >= 5) {
            textView.text = "block's total: ${blockTotalTime.average()}, block's average: ${blockAverageTime.average()}, block's onCreate: ${blockOnCreateTotalTime.average()}; " +
                    "widget's total: ${widgetTotalTime.average()}, widget's average: ${widgetAverageTime.average()}, widget's onCreate: ${widgetOnCreateTotalTime.average()}; " +
                    "fragment's total: ${fragmentTotalTime.average()}, fragment's average: ${fragmentAverageTime.average()}, fragment's onCreate: ${fragmentOnCreateTotalTime.average()}; "
            return
        }
        textView.postDelayed({
            if (flag == 0) {
                startActivityForResult(Intent(this, MainActivity5::class.java), REQUEST_CODE)
            } else if (flag == 1) {
                startActivityForResult(Intent(this, MainActivity6::class.java), REQUEST_CODE)
            } else {
                startActivityForResult(Intent(this, MainActivity7::class.java), REQUEST_CODE)
                startCount++
            }
        }, DELAYED_TIME)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE && resultCode == Activity.RESULT_OK && data != null) {
            val totalTime = data.getLongExtra(TOTAL_TIME_KEY, 0L)
            val averageTime = data.getLongExtra(AVERAGE_TIME_KEY, 0L)
            val onCreateTotalTime = data.getLongExtra(ONCREATE_TOTAL_TIME_KEY, 0L)
            val msg = if (flag == 0) {
                blockTotalTime.add(totalTime)
                blockAverageTime.add(averageTime)
                blockOnCreateTotalTime.add(onCreateTotalTime)
                "block"
            } else if (flag == 1) {
                widgetTotalTime.add(totalTime)
                widgetAverageTime.add(averageTime)
                widgetOnCreateTotalTime.add(onCreateTotalTime)
                "widget"
            } else {
                fragmentTotalTime.add(totalTime)
                fragmentAverageTime.add(averageTime)
                fragmentOnCreateTotalTime.add(onCreateTotalTime)
                "fragment"
            }
            showToast("total time of $msg: $totalTime, average time of $msg: $averageTime, onCreate time of $msg: $onCreateTotalTime")
            flag = (flag + 1) % 3
            testBlockAndWidgetAndFragment((application as ApplicationTest).textView)
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onDestroy() {
        super.onDestroy()
        (application as ApplicationTest).textView.apply {
            setOnClickListener(null)
            text = "app text"
        }
    }

    companion object {
        var MODE = false
        const val SIZE = 300
        const val ASYNC = true
        const val DELAYED_TIME = 3000L
        const val REQUEST_CODE = 33333
        const val TOTAL_TIME_KEY = "TOTAL_TIME_KEY"
        const val ONCREATE_TOTAL_TIME_KEY = "SYNC_TOTAL_TIME_KEY"
        const val AVERAGE_TIME_KEY = "AVERAGE_TIME_KEY"
    }
}
